File: csharp/Core/Storage.cs
using System.Text.Json;

namespace Aurelion.Connector.Core;

public static class Storage
{
    private const string DefaultBase = ".lake";

    public static Dictionary<string, string> WriteRecords(
        string datasetType,
        IEnumerable<Dictionary<string, object?>> records,
        string? correlationId = null)
    {
        var safe = SanitizeDatasetType(datasetType);
        var basePath = ResolveBasePath();
        var key = Guid.NewGuid().ToString();
        var dir = Path.Combine(basePath, safe);
        Directory.CreateDirectory(dir);
        var filePath = Path.Combine(dir, $"{key}.jsonl");

        using (var writer = new StreamWriter(filePath, append: false, System.Text.Encoding.UTF8))
        {
            var opts = new JsonSerializerOptions { WriteIndented = false };
            foreach (var record in records)
            {
                writer.WriteLine(JsonSerializer.Serialize(record, opts));
            }
        }

        var cidLabel = string.IsNullOrEmpty(correlationId)
            ? "[correlation_id: n/a]"
            : $"[correlation_id: {correlationId}]";
        Console.WriteLine($"{DateTimeOffset.UtcNow:o} {cidLabel} datalake write {safe}/{key}");

        return new Dictionary<string, string>
        {
            ["provider"] = "file",
            ["storage_key"] = $"{safe}/{key}",
        };
    }

    private static string ResolveBasePath()
    {
        var env = Environment.GetEnvironmentVariable("AURELION_LAKE_PATH");
        return string.IsNullOrEmpty(env) ? Path.Combine(Directory.GetCurrentDirectory(), DefaultBase) : env;
    }

    private static string SanitizeDatasetType(string datasetType)
    {
        if (datasetType.Contains("..") || datasetType.Contains('/') || datasetType.Contains('\\'))
            throw new ArgumentException($"Invalid dataset_type: '{datasetType}'");
        return datasetType;
    }
}

File: java/src/main/java/com/aurelion/connector/core/Storage.java
package com.aurelion.connector.core;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;

public final class Storage {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final String DEFAULT_BASE = ".lake";

    private Storage() {}

    public static Map<String, String> writeRecords(String datasetType,
                                                     Iterable<Map<String, Object>> records,
                                                     String correlationId) {
        String safe = sanitizeDatasetType(datasetType);
        Path base = resolveBasePath();
        String key = UUID.randomUUID().toString();
        Path dir = base.resolve(safe);

        try {
            Files.createDirectories(dir);
            Path filePath = dir.resolve(key + ".jsonl");
            try (BufferedWriter writer = Files.newBufferedWriter(filePath, StandardCharsets.UTF_8)) {
                for (Map<String, Object> record : records) {
                    writer.write(MAPPER.writeValueAsString(record));
                    writer.write('\n');
                }
            }
        } catch (IOException e) {
            throw new RuntimeException("Failed to write records", e);
        }

        String cidLabel = (correlationId == null || correlationId.isEmpty())
                ? "[correlation_id: n/a]"
                : "[correlation_id: " + correlationId + "]";
        System.out.println(Instant.now() + " " + cidLabel + " datalake write " + safe + "/" + key);

        return Map.of("provider", "file", "storage_key", safe + "/" + key);
    }

    private static Path resolveBasePath() {
        String env = System.getenv("AURELION_LAKE_PATH");
        if (env != null && !env.isBlank()) return Paths.get(env);
        return Paths.get(System.getProperty("user.dir"), DEFAULT_BASE);
    }

    private static String sanitizeDatasetType(String datasetType) {
        if (datasetType.contains("..") || datasetType.contains("/") || datasetType.contains("\\")) {
            throw new IllegalArgumentException("Invalid dataset_type: '" + datasetType + "'");
        }
        return datasetType;
    }
}
